packages/teams-js/src/public/video.ts

```diff
@@ -116,25 +207,124 @@ export namespace video {
+   * The host will choose the appropriate callback based on the host's capability.
+   *
+   * @example
+   * ```typescript
+   * video.registerForVideoFrame({
+   *   mediaStreamCallback: async (receivedVideoFrame) => {
+   *     const originalFrame = receivedVideoFrame.videoFrame as VideoFrame;
+   *     try {
+   *       const processedFrame = await processFrame(originalFrame);
+   *       return processedFrame;
+   *     } catch (e) {
+   *       throw e;
+   *     }
+   *   },
+   *   sharedFrameCallback: (
+   *     frame: VideoFrameData,
+   *     notifyVideoFrameProcessed: notifyVideoFrameProcessedFunctionType,
+   *     notifyError: notifyErrorFunctionType
+   *     ) => {
+   *       try {
+   *         processFrameInplace(frame);
+   *         notifyVideoFrameProcessed();
+   *       } catch (e) {
+   *         notifyError(e);
+   *       }
+   *     },
+   *   config: {
+   *     format: video.VideoPixelFormat.NV12,
+   *   }
+   * });
+   * ```
    */
-  export function registerForVideoFrame(frameCallback: VideoFrameCallback, config: VideoFrameConfig): void {
+  export function registerForVideoFrame(parameters: RegisterForVideoFrameParameters): void {
     ensureInitialized(runtime, FrameContexts.sidePanel);
     if (!isSupported()) {
       throw errorNotSupportedOnPlatform;
     }
+    if (!parameters.mediaStreamCallback || !parameters.sharedFrameCallback) {
+      throw new Error('Both mediaStreamCallback and sharedFrameCallback must be provided');
+    }
+    if (doesSupportMediaStream()) {
+      registerForMediaStream(parameters.mediaStreamCallback);
+    } else if (doesSupportSharedFrame()) {
+      registerForSharedFrame(parameters.sharedFrameCallback, parameters.config);
+    } else {
+      // should not happen if isSupported() is true
+      throw errorNotSupportedOnPlatform;
+    }
+  }
+
+  function doesSupportMediaStream(): boolean {
+    return (
+      ensureInitialized(runtime, FrameContexts.sidePanel) &&
+      isTextureStreamAvailable() &&
+      !!runtime.supports.video?.mediaStream
+    );
+  }
+
+  function isTextureStreamAvailable(): boolean {
+    return (
+      !inServerSideRenderingEnvironment() &&
+      !!(window['chrome']?.webview?.getTextureStream && window['chrome']?.webview?.registerTextureStream)
+    );
+  }
+
+  function doesSupportSharedFrame(): boolean {
+    return ensureInitialized(runtime, FrameContexts.sidePanel) && !!runtime.supports.video?.sharedFrame;
+  }
+
+  function registerForMediaStream(mediaStreamCallback: MediaStreamCallback): void {
+    ensureInitialized(runtime, FrameContexts.sidePanel);
+    if (!isSupported() || !doesSupportMediaStream()) {
+      throw errorNotSupportedOnPlatform;
+    }
+
+    registerHandler('video.startVideoExtensibilityVideoStream', async (mediaStreamInfo: MediaStreamInfo) => {
+      // when a new streamId is ready:
+      const { streamId } = mediaStreamInfo;
+      const generator = await processMediaStream(streamId, mediaStreamCallback, notifyError);
```

Comment:
In videoEx, the `registerForMediaStream` function would be very similar to this, but have its own implementation of `processMediaStream`, so I wonder if we could make this function shared

Reply:
let's do it when we are going to update `videoEx.registerForVideoFrame`